core/dataset_next.cpp

```diff
@@ -0,0 +1,602 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#include <ostream>
+
+#include "dataset_next.h"
+#include "dataset.h"
+#include "except.h"
+
+namespace scipp::core::next {
+
+template <class T>
+std::pair<const Variable *, Variable *> makeProxyItem(T *variable) {
+  if constexpr (std::is_const_v<T>)
+    return {variable, nullptr};
+  else
+    return {variable, variable};
+}
+
+template <class Key, class T1> auto makeProxyItems(T1 &coords) {
+  std::map<Key, std::pair<const Variable *, Variable *>> items;
+  for (auto &item : coords)
+    items.emplace(item.first, makeProxyItem(&item.second));
+  return items;
+}
+
+template <class Key, class T1, class T2 = void>
+auto makeProxyItems(const Dimensions &dims, T1 &coords,
+                    const Dim sparseDim = Dim::Invalid, T2 *sparse = nullptr) {
+  std::map<Key, std::pair<const Variable *, Variable *>> items;
+  for (auto &item : coords) {
+    // We preserve only items that are part of the space spanned by the
+    // provided parent dimensions. Note the use of std::any_of (not
+    // std::all_of): At this point there may still be extra dimensions in item,
+    // but they will be sliced out. Maybe a better implementation would be to
+    // slice the coords first? That would also eliminate a potential loophole
+    // for multi-dimensional coordinates.
+    auto contained = [&dims](const auto item) {
+      const auto &coordDims = item.second.dims();
+      if constexpr (std::is_same_v<Key, Dim>)
+        return coordDims.empty() || dims.contains(item.first);
```

Comment:
Minor refactor: lambda -> named function. Would make slightly more readable esp given nested and multiple returns relating to the lambda.